ports/stm32/storage.c

```diff
@@ -341,7 +341,11 @@ STATIC mp_obj_t pyb_flash_readblocks(size_t n_args, const mp_obj_t *args) {
         // Extended block read on a sub-section of the flash storage
         uint32_t offset = mp_obj_get_int(args[3]);
         block_num += self->start / PYB_FLASH_NATIVE_BLOCK_SIZE;
-        ret = MICROPY_HW_BDEV_READBLOCKS_EXT(bufinfo.buf, block_num, offset, bufinfo.len);
+        if (block_num >= storage_get_block_count()) {
```

Comment:
I think storage_get_block_count() will use different units (512 block size) to what MICROPY_HW_BDEV_READBLOCKS_EXT uses (flash erase size).

Reply:
Ah yes, that would make sense. It certainly could, if it's auto-configured to the flash block size instead. This will need a little more thought...

Reply:
You might be able to use `self->len` here to calculate the correct maximum number of blocks, because the block size should be `PYB_FLASH_NATIVE_BLOCK_SIZE`.

Reply:
So, as mentioned `storage_write_blocks()` and `storage_read_blocks()` are already protected for valid blocks, leaving just the 
`MICROPY_HW_BDEV_READBLOCKS_EXT()` and `MICROPY_HW_BDEV_WRITEBLOCKS_EXT()` functions needing protection, as this PR is currently targeting.

Depending on flash configuration, the write function can be defined as `spi_bdev_writeblocks_raw()` or `flash_bdev_writeblocks_ext()` (with matching function pair for the read), done at the top of `stm32/storage.c`. Next to each of these is a matching definition for `PYB_FLASH_NATIVE_BLOCK_SIZE`

* `spi_bdev_writeblocks_raw()` converts the passed in block number to address with `block_num * MP_SPIFLASH_ERASE_BLOCK_SIZE + block_offset`
* `flash_bdev_writeblocks_ext()` converts block to address with `flashbdev.c/convert_block_to_flash_addr()` which uses `FLASH_BLOCK_SIZE`

These do correspond to the `PYB_FLASH_NATIVE_BLOCK_SIZE` definitions, so yes I agree that this should be safe to use.

However, `pyb_flash_ioctl()` makes it look like this may not always be the case? 
``` c
        case MP_BLOCKDEV_IOCTL_INIT: {
            mp_int_t ret = 0;
            storage_init();
            if (mp_obj_get_int(arg_in) == 1) {
                // Will be using extended block protocol
                if (self == &pyb_flash_obj) {
                    ret = -1;
                #if defined(SPIFLASH)
                } else {
                    // Switch to use native block size of SPI flash
                    self->use_native_block_size = true;
                #endif
                }
```
This is then used similarly when it returns block count
``` c
        case MP_BLOCKDEV_IOCTL_BLOCK_COUNT: {
            mp_int_t n;
            if (self == &pyb_flash_obj) {
                // Get true size
                n = storage_get_block_count();
            #if defined(SPIFLASH)
            } else if (self->use_native_block_size) {
                n = self->len / PYB_FLASH_NATIVE_BLOCK_SIZE;
            #endif
            } else {
                n = self->len / FLASH_BLOCK_SIZE;
            }
```
Using `FLASH_BLOCK_SIZE` if use_native_block_size  wasn't enabled during runtime init. This is different to the `pyb_flash_readblocks()` function where `block_num += self->start / PYB_FLASH_NATIVE_BLOCK_SIZE;` is always using `
PYB_FLASH_NATIVE_BLOCK_SIZE`.

This to me looks like a separate bug? `pyb_flash_readblocks()` and `pyb_flash_writeblocks()` always use PYB_FLASH_NATIVE_BLOCK_SIZE for offsetting the start address, however if the runtime `self->use_native_block_size` it looks like `FLASH_BLOCK_SIZE` should be used?